src/Api.cpp

```diff
@@ -1,41 +1,108 @@
+
+		heartbeat_interval = -1;
+		connected = false;
+
+		webSocket.setOnMessageCallback([&](const ix::WebSocketMessagePtr& msg)
+			{
+				if (msg->type == ix::WebSocketMessageType::Message)
+				{
+					if (dev) {std::cout << "RESPONSE: " << msg->str << "\n\n";}
+					json res = json::parse(msg->str);
+					if (res["s"].is_number())
+						lastS = res["s"];
+					int op = std::stoi(res["op"].dump());
+					switch (op) {
+					case 10:
+						heartbeat_interval = std::stoi(res["d"]["heartbeat_interval"].dump());
+						if (!connected) {
+							connected = true;
+							json heartbeat_p = {
+								{"op", 1},
+								{"d", nullptr}
+							};
+							if(dev){ std::cout << "SENDING HEARTBEAT: " << heartbeat_p.dump() << "\n\n"; }
+							webSocket.send(heartbeat_p.dump());
+						}
+						sendID();
+						break;
+					case 0:
+						std::string t = res["t"];
+						if (t == "READY") {
+							if(dev){ std::cout << "USER INFO: " << res["d"]["user"] << "\n\n"; }
+							user.initialize(res["d"]["user"]);
+							if (onReady)
+								onReady();
+						}
+						else if (t == "MESSAGE_CREATE") {
+							dpp::Message message(res["d"]);
+							if (onMessage)
+								onMessage(message);
+						}
+						break;
+					}
+				}
+				else if (msg->type == ix::WebSocketMessageType::Open)
+				{
+					std::cout << "====================\nConnection established\n====================\n\n";
+				}
+			}
 		);
-		return json::parse(res.text);
+
+		webSocket.start();
+
+		int count = 0;
+		while (true)
+		{
+			count += 200;
+			if (connected && heartbeat_interval != -1 && count >= heartbeat_interval) {
+				count = 0;
+				json payload = {
+					{"op", 1},
+					{"d", lastS}
+				};
+				if(dev){ std::cout << "SENDING HEARTBEAT: " << payload.dump() << "\n\n"; }
+				webSocket.send(payload.dump());
+			}
+			std::this_thread::sleep_for(std::chrono::milliseconds(200));
+		}
 	}
-	json put(const std::string& path) {
-		const std::string URL = baseURL + path;
-		std::cout << "URL: " << URL << std::endl;
-		cpr::Response res = cpr::Put(
-			cpr::Url{ URL },
-			cpr::Header{ {"authorization", auth}, {"Content-length", "0"} }
-		);
-		std::cout << "Response: " << res.text << std::endl;
-		json result;
-		if (res.text.length() != 0)
-			result = json::parse(res.text);
-		else
-			result = json::parse("{}");
-		return result;
+	void Client::sendID() {
+		json props = { {"$os", "linux"}, {"$browser", "discord_cpp"}, {"$device", "discord_cpp"} };
+		json id = {
+			{"op", 2},
+			{"d", {
+				{"token", token}, {"properties", props}, {"intents", 513}
+			}}
+		};
+		if(dev){ std::cout << "SENDING: " << id.dump() << "\n\n"; }
+		webSocket.send(id.dump());
+	}
+	void Client::send(const std::string message, const std::string channel_id) {
+		const std::string path = "/channels/" + channel_id + "/messages";
+		json body = { {"content", message} };
+		json res = Api::post(path, cpr::Body{ body.dump() });
+		if(dev){ std::cout << res << std::endl; }
 	}
 }
```

Comment:
Don't fully understand what's going on here. Did you mean to add this changes to Client.cpp instead of Api.cpp?

Reply:
No worries